src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2210,7 +2210,10 @@ exprt c_typecheck_baset::do_special_functions(
     isnan_exprt isnan_expr(expr.arguments().front());
     isnan_expr.add_source_location()=source_location;
 
-    return isnan_expr;
+    if(expr.type()!=isnan_expr.type())
+      return typecast_exprt(isnan_expr, expr.type());
+    else
+      return isnan_expr;
```

Comment:
This is by far not the only place where we are doing the "if the types don't match, add a typecast" dance. Could 1) `exprt::make_typecast` be removed (it really doesn't do anything interesting and 2) something like `add_cast_if_needed` be added to expr_util.h?

Reply:
ok, done, added typecast_exprt::conditional_cast
exprt::make_typecast has been deprecated for quite a while.